### PR TITLE
Move common pitfall documentation to Airflow docs

### LINKED ISSUE #10180: Review and move Common Pitfalls Documentation to Airflow Docs
**Description**

The old documentation for [Common Pitfalls](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=62694614) should be reviewed and incorporated in [Airflow Docs](https://github.com/apache/airflow/tree/master/docs)

**Use case / motivation**

To remove documentation duplication and clean it up.

**Related Issues**

#10178

### PR DESCRIPTION
**Pull request by wolfier**
---
Reviewed Common Pitfall page and migrated to Airflow documentation. I also added more common pitfalls.

closes: #10180

### CODE REVIEW

File: docs/apache-airflow/common-pitfall.rst [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+.. code-block:: python
+
+        bo = BashOperator(
+            task_id='my_script',
+            bash_command="/usr/local/airflow/include/test.sh",
+            dag=dag
+        )
+
+If you wish to directly executed a bash script, you need to add a space after the script name to prevent Airflow from
+rendering the template using Jinja.
```
Comment (eladkal) [inline docs/apache-airflow/common-pitfall.rst]: This is duplicate content. It's explained in https://github.com/apache/airflow/blob/master/docs/apache-airflow/howto/operator/bash.rst#jinja-template-not-found

You can create a ref link to `bash.rst` instead.

File: docs/apache-airflow/best-practices.rst [hunk 2]
```diff
@@ -54,6 +59,12 @@ Some of the ways you can avoid producing a different result -
     You should define repetitive parameters such as ``connection_id`` or S3 paths in ``default_args`` rather than declaring them for each task.
     The ``default_args`` help to avoid mistakes such as typographical errors.
 
+Creating a custom Operator
+---------------------------
+
+When implementing custom operators, do not make any expensive expensive operations in their ``__init__``. They are
+going to be instantiated once per scheduler run per task using them, and making database calls can significantly slow
+down scheduling and waste resources.
```
Comment (eladkal) [inline docs/apache-airflow/best-practices.rst]: This is probably something we need to mention in https://github.com/apache/airflow/blob/master/docs/apache-airflow/howto/custom-operator.rst#creating-a-custom-operator and create a ref link there.

**Review by eladkal — CHANGES_REQUESTED** [review]
Thanks @wolfier for you contribution!

I'm wondering about `faq.rst` <-> `common-pitfall.rst` Do we really need both?
Isn't a frequently asked question by definition a pitfall?

File: docs/apache-airflow/best-practices.rst [hunk 3]
```diff
@@ -54,6 +59,12 @@ Some of the ways you can avoid producing a different result -
     You should define repetitive parameters such as ``connection_id`` or S3 paths in ``default_args`` rather than declaring them for each task.
     The ``default_args`` help to avoid mistakes such as typographical errors.
 
+Creating a custom Operator
+---------------------------
+
+When implementing custom operators, do not make any expensive expensive operations in their ``__init__``. They are
```
Comment (marcosmarxm) [inline docs/apache-airflow/best-practices.rst]:
```suggestion
When implementing custom operators, do not make any expensive operations in their ``__init__``.
```
Small typo

File: docs/apache-airflow/common-pitfall.rst [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+Sensing tasks from other DAGs
+-----------------------------
+
+When waiting on another task's result, it is recommended to raise the priority of the tasks it will be waiting for.
+Otherwise, scheduling can get deadlocked, with the waiting task taking up all available slots, leaving no slots for the
+execution of the tasks they are waiting for. It is also a good idea for the waiting task to have a later scheduler
+interval than the task it is waiting on.
+
+`ExternalTaskSensor` is an example of a task sensing on the state of a task in another DAG that could potentially run
```
Comment (marcosmarxm) [inline docs/apache-airflow/common-pitfall.rst]:
```suggestion
``ExternalTaskSensor`` is an example of a task sensing on the state of a task in another DAG that could potentially run
```
double ` for code in rst

File: docs/apache-airflow/common-pitfall.rst [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+  The maximum number of task instances to be allowed to run concurrently
+  per DAG. To calculate the number of tasks that is running concurrently
+  for a DAG, add up the number of running tasks for all DAG runs of the DAG.
+  This is configurable at the DAG level with ``concurrency``.
+
+- :ref:`config:core__max_active_runs_per_dag`
+
+  The maximum number of active DAG runs per DAG. The scheduler will not
+  create more DAG runs if it reaches the limit. This is configurable at
+  the DAG level with ``max_active_runs``.
```
Comment (marcosmarxm) [inline docs/apache-airflow/common-pitfall.rst]: These appear to duplicate the content in [here](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html)
why not update configurations to be more descriptive and reference here?


There some content in [FAQ How can my airflow dag run faster](https://airflow.apache.org/docs/apache-airflow/stable/faq.html#how-can-my-airflow-dag-run-faster) that are correlated with this too, WDYT?

**Comment by marcosmarxm** [comment]
> Thanks @wolfier for you contribution!
> 
> I'm wondering about `faq.rst` <-> `common-pitfall.rst` Do we really need both?
> Isn't a frequently asked question by definition a pitfall?

Maybe improve FAQ to separate questions into bigger topics: Scheduler, Operators/Hooks, Configurations.

File: docs/apache-airflow/common-pitfall.rst (hunk 1 above)
Comment (wolfier) [reply docs/apache-airflow/common-pitfall.rst]: Ahh! Yeah, I will add a reference to `bash.rst`.

On the same note, Should I move the the relative path interaction and the interaction between `template_fields` and `template_ext` into `faq.rst` since that issue is not limited to only the BashOperator. I can reference `faq.rst` in `bash.rst` instead.

File: docs/apache-airflow/best-practices.rst (hunk 2 above)
Comment (wolfier) [reply docs/apache-airflow/best-practices.rst]: Got it!

File: docs/apache-airflow/best-practices.rst (hunk 3 above)
Comment (wolfier) [reply docs/apache-airflow/best-practices.rst]: Thanks for catching that!

File: docs/apache-airflow/common-pitfall.rst (hunk 4 above)
Comment (wolfier) [reply docs/apache-airflow/common-pitfall.rst]: Good catch!

File: docs/apache-airflow/common-pitfall.rst (hunk 5 above)
Comment (wolfier) [reply docs/apache-airflow/common-pitfall.rst]: Yeah that makes sense! I will update the configuration description and update that section with the links to those airflow configurations.

**Comment by wolfier** [comment]
@marcosmarxm I think I'll save that for another PR. I feel like it's out of scope.

I mainly want to make sure all the content in [Common Pitfall ](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=62694614) is somewhere in the Airflow documentation and that they are still valid.

File: docs/apache-airflow/faq.rst [hunk 6]
```diff
@@ -159,72 +210,206 @@ simple dictionary.
...
 
+
+Task execution interactions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Why does a sensor never complete?
+---------------------------------
+
+When waiting on another task's result, it is recommended to raise the priority of the tasks it will be waiting for.
+Otherwise, scheduling can get deadlocked, with the waiting task taking up all available slots, leaving no slots for the
```
Comment (eladkal) [inline docs/apache-airflow/faq.rst]: What do you mean by raise priority of the tasks?

File: docs/apache-airflow/faq.rst [hunk 7]
```diff
@@ -159,72 +210,206 @@ simple dictionary.
...
+
+``ExternalTaskSensor`` is an example of a task sensing on the state of a task in another DAG that could potentially run
+into this deadlock issue.
+
+
+What does TemplateNotFound mean?
+---------------------------------
+
+TemplateNotFound errors are usually due to misalignment with user expectations when passing special values to Operator
+that trigger Jinja templating. A common occurrence is with :ref:`BashOperators<howto/operator:BashOperator>`.
```
Comment (eladkal) [inline docs/apache-airflow/faq.rst]:
```suggestion
TemplateNotFound errors are usually due to misalignment with user expectations when passing path to operator
that trigger Jinja templating. A common occurrence is with :ref:`BashOperators<howto/operator:BashOperator>`.
```

**Review by eladkal — CHANGES_REQUESTED** [review]

File: docs/apache-airflow/faq.rst (hunk 6 above)
Comment (wolfier) [reply docs/apache-airflow/faq.rst]: Ah! raising the priority_weight. I will include that to be more specific.

File: docs/apache-airflow/faq.rst (hunk 7 above)
Comment (wolfier) [reply docs/apache-airflow/faq.rst]: Good suggestion! I will put that in.

File: docs/apache-airflow/faq.rst (hunk 6 above)
Comment (eladkal) [reply docs/apache-airflow/faq.rst]: I'm not that comfortable with that recommendation.
priority weight should be used because you want to provide a task priority not because you want to "help" the scheduler to avoid deadlock.
Also the paragraph doesn't consider the mode of the sensor. If you are running in `mode='poke'` the recommendation isn't relevant.

Comment (wolfier) [reply docs/apache-airflow/faq.rst]: I will remove this faq for now.

**Comment by wolfier** [comment]
@eladkal and @marcosmarxm and @kaxil, please review.

File: docs/apache-airflow/concepts/pools.rst [hunk 8]
```diff
@@ -33,14 +37,16 @@ Tasks can then be associated with one of the existing pools by using the ``pool`
...
 
-The ``pool`` parameter can be used in conjunction with the ``priority_weight`` parameter to define priorities in the queue, and which tasks get executed first as slots open up in the pool.
-
-The default ``priority_weight`` is ``1``, and can be bumped to any number. When sorting the queue to evaluate which task should be executed next, we use the ``priority_weight``, summed up with all of the ``priority_weight`` values from tasks downstream from this task; the highest summed value wins. Thus, you can bump a specific important task, and the whole path to that task gets prioritized accordingly.
 
-Tasks will be scheduled as usual while the slots fill up. Once capacity is reached, runnable tasks get queued and their state will show as such in the UI. As slots free up, queued tasks start running based on the ``priority_weight`` (of the task and its descendants).
+Tasks will be scheduled as usual while the slots fill up. Once capacity is reached, runnable tasks get queued and their
+state will show as such in the UI. As slots free up, queued tasks start running based on the
+:ref:`concepts:priority-weight`.
+(of the task and its descendants).
```
Comment (eladkal) [inline docs/apache-airflow/concepts/pools.rst]: I think it's worth mentioning that the number of slots occupied by a task can be configured by `pool_slots`

File: docs/apache-airflow/concepts/priority-weight.rst [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    under the License.
+
+.. _concepts:priority-weight:
+
+Priority Weights
+================
+
+``priority_weight`` defines priorities in the executor queue. The default ``priority_weight`` is ``1``, and can be
+bumped to any number. When sorting the executor queue to evaluate which task should be executed next, we calculate the
+true ``priority_weight``, which is summed up with all of the ``priority_weight`` values from its downstream tasks.
```
Comment (eladkal) [inline docs/apache-airflow/concepts/priority-weight.rst]: This is true only if `weight_rule='downstream'`. I think we should explain also about weight_rule parameter in this section.

File: docs/apache-airflow/concepts/priority-weight.rst [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+You can use this to bump a specific important task and the whole path to that task gets prioritized accordingly.
+
+The ``priority_weight`` parameter can be used in conjunction with :ref:`concepts:pool`.
+
+Tasks will be scheduled as usual while the slots fill up. Once capacity is reached, runnable tasks get queued and their
+state will show as such in the UI. As slots free up, queued tasks start running based on the
+``priority_weight`` (of the task and its descendants).
```
Comment (eladkal) [inline docs/apache-airflow/concepts/priority-weight.rst]: This is the exact text you wrote in lines 41-44 of pools.rst.
I think this text isn't relevant here.

File: docs/apache-airflow/faq.rst [hunk 11]
```diff
@@ -159,72 +215,194 @@ simple dictionary.
         other_dag_id = f'bar_{i}'
         globals()[other_dag_id] = create_dag(other_dag_id)
 
-What are all the ``airflow tasks run`` commands in my process list?
--------------------------------------------------------------------
+Even though Airflow supports multiple DAG definition per python file, dynamically generated or else, it is not
+recommended as Airflow would like better isolation between DAGs from a fault and deployment perspective and multiple
+DAGs in the same file goes against that.
```
Comment (eladkal) [inline docs/apache-airflow/faq.rst]: Can you clarify a bit about this? This feels a bit harsh I read this paragraph several times I'm not sure I fully understand it. This paragraph is under the question: "How to create DAGs dynamically?" Are you recommending against dynamic DAGs? (as their base code is in one file)

File: docs/apache-airflow/faq.rst [hunk 12]
```diff
@@ -159,72 +215,194 @@ simple dictionary.
...
+    )
+
+    child = PythonOperator(
+        task_id='b',
+        python_callable=lambda: 1,
+        trigger_rule=TriggerRule.ALL_FAILED,
+        dag=dag
+    )
+
+    parent >> child
```
Comment (eladkal) [inline docs/apache-airflow/faq.rst]: lets use Airflow 2 syntax with dag & task decorators.

**Review by eladkal — DISMISSED** [review]
Looks good. left some comments

File: docs/apache-airflow/concepts/pools.rst (hunk 8 above)
Comment (wolfier) [reply docs/apache-airflow/concepts/pools.rst]: I didn't know about this feature! Will add to the doc.

File: docs/apache-airflow/concepts/priority-weight.rst (hunk 10 above)
Comment (wolfier) [reply docs/apache-airflow/concepts/priority-weight.rst]: Oops must have copied it by accident when refactoring to master. Will remove.

File: docs/apache-airflow/concepts/priority-weight.rst (hunk 9 above)
Comment (wolfier) [reply docs/apache-airflow/concepts/priority-weight.rst]: Good idea! I could only find documentation mentioning weight_rule from the [BaseOperator docstring](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/index.html?highlight=weight_rule#airflow.models.BaseOperator).

File: docs/apache-airflow/faq.rst (hunk 11 above)
Comment (wolfier) [reply docs/apache-airflow/faq.rst]: The line came verbatim from the old [Common Pitfall](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=62694614) page.

> We do support more than one DAG definition per python file, but it is not recommended as we would like better isolation between DAGs from a fault and deployment perspective and multiple DAGs in the same file goes against that. For now, make sure that the dag object is in the global namespace : you can use the globals dict as in globals()[dag_id] = DAG(...)

I think the original author wanted to affirm that Airflow does allow a user to define multiple DAGs in a file whether they are statically defining those DAGs or in a loop of some sorts and using the globals symbol table to store the definitions. However, having multiple DAG definition goes against Airflow's mindset/ideal of isolation between DAGs from a fault and deployment perspective.

To sum it up, yes, you can create dynamic DAGs but Airflow does not recommend it. 

Unless we do recommend it (?)

Comment (wolfier) [reply docs/apache-airflow/faq.rst]: Or should that snippet be placed elsewhere in the documentation?

File: docs/apache-airflow/faq.rst (hunk 12 above)
Comment (wolfier) [reply docs/apache-airflow/faq.rst]: Sounds good! Will convert to task decorator.

File: docs/apache-airflow/faq.rst (hunk 11 above)
Comment (wolfier) [reply docs/apache-airflow/faq.rst]: @eladkal

Comment (eladkal) [reply docs/apache-airflow/faq.rst]: >To sum it up, yes, you can create dynamic DAGs but Airflow does not recommend it.

:thinking:
@vikramkoka @kaxil @ashb @potiuk Can you share your point of view on this ?

File: docs/apache-airflow/faq.rst [hunk 13]
```diff
@@ -159,72 +215,205 @@ simple dictionary.
...
   ``execution_date`` and ``schedule_interval``. If you set ``schedule_interval`` as ``None`` or ``@once``,
   the ``next_ds``, ``next_ds_nodash``, ``prev_ds``, ``prev_ds_nodash`` values will be set to ``None``.
 - When manually triggering DAG, the schedule will be ignored, and ``prev_ds == next_ds == ds``
 
+
+Task execution interactions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+What does TemplateNotFound mean?
+---------------------------------
```
Comment (kaxil) [inline docs/apache-airflow/faq.rst]:
```suggestion
What does ``TemplateNotFound`` mean?
-------------------------------------
```

File: docs/apache-airflow/faq.rst [hunk 14]
```diff
@@ -159,72 +215,205 @@ simple dictionary.
...
+
+TemplateNotFound errors are usually due to misalignment with user expectations when passing path to operator
```
Comment (kaxil) [inline docs/apache-airflow/faq.rst]:
```suggestion
``TemplateNotFound`` errors are usually due to misalignment with user expectations when passing path to operator
```

File: docs/apache-airflow/faq.rst [hunk 15]
```diff
@@ -159,72 +215,205 @@ simple dictionary.
...
+
+You will experience UnicodeDecodeError thrown by WTForms templating and other Airflow modules like below.
+
+.. code-block:: text
+
+   'ascii' codec can't decode byte 0xae in position 506: ordinal not in range(128)
+
+
+How to fix Exception: Global variable explicit_defaults_for_timestamp needs to be on (1)?
+-----------------------------------------------------------------------------------------
```
Comment (kaxil) [inline docs/apache-airflow/faq.rst]:
```suggestion
How to fix Exception: Global variable ``explicit_defaults_for_timestamp`` needs to be on (1)?
---------------------------------------------------------------------------------------------
```

File: docs/apache-airflow/faq.rst [hunk 16]
```diff
@@ -159,72 +215,205 @@ simple dictionary.
...
+----------------------------------------------------------
+
+If you intend to use extended ASCII or Unicode characters in Airflow, you have to provide a proper connection string to
+the MySQL database since they define charset explicitly.
+
+.. code-block:: text
+
+    sql_alchemy_conn = mysql://airflow@localhost:3306/airflow?charset=utf8
+
+You will experience UnicodeDecodeError thrown by WTForms templating and other Airflow modules like below.
```
Comment (kaxil) [inline docs/apache-airflow/faq.rst]:
```suggestion
You will experience ``UnicodeDecodeError`` thrown by ``WTForms`` templating and other Airflow modules like below.
```

File: docs/apache-airflow/faq.rst [hunk 17]
```diff
@@ -159,72 +215,205 @@ simple dictionary.
...
 issue recurs.
+
+
+MySQL and MySQL variant Databases
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+What does "MySQL Server has gone away" mean?
+--------------------------------------------
+
+You may occasionally experience OperationalError with the message "MySQL Server has gone away". This is due to the
```
Comment (kaxil) [inline docs/apache-airflow/faq.rst]:
```suggestion
You may occasionally experience ``OperationalError`` with the message "MySQL Server has gone away". This is due to the
```

File: docs/apache-airflow/faq.rst [hunk 18]
```diff
@@ -159,72 +215,205 @@ simple dictionary.
...
+  ``airflow tasks run --local`` command. If using Celery, this means it puts a
+  command in the queue for it to run remotely on the worker. If using
+  LocalExecutor, that translates into running it in a subprocess pool.
+- Local ``airflow tasks run --local``: starts an ``airflow tasks run --raw``
+  command (described below) as a subprocess and is in charge of
+  emitting heartbeats, listening for external kill signals
+  and ensures some cleanup takes place if the subprocess fails.
+- Raw ``airflow tasks run --raw`` runs the actual operator's execute method and
+  performs the actual work.
+
```
Comment (kaxil) [inline docs/apache-airflow/faq.rst]: Let's move this to separate PR plz

```suggestion
```

File: docs/apache-airflow/faq.rst [hunk 19]
```diff
@@ -159,72 +215,205 @@ simple dictionary.
...
+        a >> b
+
+    dag = my_dag()
+
+See :ref:`concepts:trigger-rules` for more information.
+
+If the tasks are not related by dependency, you will need to :ref:`build a custom Operator<custom_operator>`.
+
+
```
Comment (kaxil) [inline docs/apache-airflow/faq.rst]:
```suggestion
```

**Review by kaxil — APPROVED** [review]
